### ISSUE TITLE
Halting while waiting for API server

### ISSUE THREAD

**Issue by rtlong** [kz8s/tack#5]
Hey, first of all, this repo is awesome. Thank you so much for sharing it!

I've been reading through everything and I think I have a handle on the issue I'm seeing, but I'm still new to K8s and I'm not sure where to begin trying to solve it. So I'm hoping for some support here, more than calling this a "bug".

Basically, I'm seeing the terraform process halt during the `initialize` step when it's doing the remote `curl http://127.0.0.1:8080/` – upon inspection, I don't see the kube API server running on any of the nodes, so it makes sense that it's stuck there. I see that the intention is to have an HA-setup of the kube API service, so I gather we're doing things differently than might be normal, here. I also see that there is a `kube-apiserver.yaml` k8s manifest, which leads me to believe that we're trying to have k8s run its own API service (as opposed to running the API service directly with systemd). The confusion I'm having now is about how that's supposed to work. I see in the kubelet logs   that it's also freaking out about not being able to talk to the API, but as far as I can tell, it's the kubelet which would be the only process aware of the k8s manifests and which should ultimately schedule the API service. 

Thanks in advance for being patient with my noobiness. :smile:

**Comment by wellsie** [comment]
hey ryan! you are welcome :) i figured the community could benefit from a terraform reference implementation...

could you share a few lines from:
- the console output around where the problem is happening ?
- journalctl on `10.0.0.10` where you see the freak out about not being able to talk to the apiserver

it seems the trendy thing to do these days is let `kubelet` run `apiserver`, `proxy`, `controller-manager` et al as pods. the `--config` switch to `kubelet` indicates a directory which is scanned for changes every 20s or so (not sure the exact timing). see `modules/etcd/cloud-config.tf`:

```
    - name: kubelet.service
      command: start
      content: |
        [Unit]
        After=docker.socket
        ConditionFileIsExecutable=/opt/bin/kubectl
        Requires=docker.socket
        [Service]
        ExecStart=/opt/bin/kubelet \
          --allow-privileged=true \
          --api-servers=http://127.0.0.1:8080 \
          --cluster-dns=10.3.0.10 \
          --cluster-domain=cluster.local \
          --config=/etc/kubernetes/manifests \
          --register-node=false
        Restart=always
        RestartSec=5
        [Install]
        WantedBy=multi-user.target
```

if you do `docker ps` on `10.0.0.10` you should see the containers that `kubelet` launched there (btw `kubelet` will keep these running should they die):

```
$ make ssh
$ ssh 10.0.0.10
$ docker ps
CONTAINER ID        IMAGE                                       COMMAND                  CREATED             STATUS              PORTS               NAMES
ca67ec671116        gcr.io/google_containers/hyperkube:v1.2.0   "/hyperkube scheduler"   6 minutes ago       Up 6 minutes                            k8s_kube-scheduler.7b10907e_kube-scheduler-ip-10-0-0-10.us-west-1.compute.internal_kube-system_1adf66daf2f3ee9cce476b5b97a8b58c_114ac564
a4eee538f5c3        gcr.io/google_containers/hyperkube:v1.2.0   "/hyperkube controlle"   6 minutes ago       Up 6 minutes                            k8s_kube-controller-manager.89f70b8f_kube-controller-manager-ip-10-0-0-10.us-west-1.compute.internal_kube-system_c79711b2d60ce63ccce896a8111c8ba3_f4299389
d07481055ccd        gcr.io/google_containers/hyperkube:v1.2.0   "/hyperkube proxy --m"   6 minutes ago       Up 6 minutes                            k8s_kube-proxy.51156269_kube-proxy-ip-10-0-0-10.us-west-1.compute.internal_kube-system_569f2def3d8e2df8825f616eba237006_319e3343
dd7c1bf2370e        gcr.io/google_containers/hyperkube:v1.2.0   "/hyperkube apiserver"   6 minutes ago       Up 6 minutes                            k8s_kube-apiserver.57925c18_kube-apiserver-ip-10-0-0-10.us-west-1.compute.internal_kube-system_adbb9b34587af41c37b3c16ab2b881a6_225abce7
f611f81b68de        gcr.io/google_containers/pause:2.0          "/pause"                 6 minutes ago       Up 6 minutes                            k8s_POD.6059dfa2_kube-scheduler-ip-10-0-0-10.us-west-1.compute.internal_kube-system_1adf66daf2f3ee9cce476b5b97a8b58c_2de18374
3369a7afc846        gcr.io/google_containers/pause:2.0          "/pause"                 6 minutes ago       Up 6 minutes                            k8s_POD.6059dfa2_kube-controller-manager-ip-10-0-0-10.us-west-1.compute.internal_kube-system_c79711b2d60ce63ccce896a8111c8ba3_3f676c9c
009ca4840acb        gcr.io/google_containers/podmaster:1.1      "/podmaster --etcd-se"   6 minutes ago       Up 6 minutes                            k8s_controller-manager-elector.433b6cc4_kube-podmaster-ip-10-0-0-10.us-west-1.compute.internal_kube-system_66f8cf4ddf87288d1bbbb588a4bc6077_d9878022
250152b90047        gcr.io/google_containers/podmaster:1.1      "/podmaster --etcd-se"   6 minutes ago       Up 6 minutes                            k8s_scheduler-elector.c6c6198_kube-podmaster-ip-10-0-0-10.us-west-1.compute.internal_kube-system_66f8cf4ddf87288d1bbbb588a4bc6077_d2c2f1dc
a627370bf6a1        gcr.io/google_containers/pause:2.0          "/pause"                 7 minutes ago       Up 7 minutes                            k8s_POD.6059dfa2_kube-proxy-ip-10-0-0-10.us-west-1.compute.internal_kube-system_569f2def3d8e2df8825f616eba237006_896345a6
653a751edddf        gcr.io/google_containers/pause:2.0          "/pause"                 7 minutes ago       Up 7 minutes                            k8s_POD.6059dfa2_kube-apiserver-ip-10-0-0-10.us-west-1.compute.internal_kube-system_adbb9b34587af41c37b3c16ab2b881a6_c7fc5770
fd4512b3b604        gcr.io/google_containers/pause:2.0          "/pause"                 7 minutes ago       Up 7 minutes                            k8s_POD.6059dfa2_kube-podmaster-ip-10-0-0-10.us-west-1.compute.internal_kube-system_66f8cf4ddf87288d1bbbb588a4bc6077_70867fa5
```

that delay you are seeing is normal - i have seen it take a minute or so for services in `/etc/kubernetes/manifest` to start. docker is pulling the containers etc.

i take it you have waited for more than a few minutes though.

lmk if this helps and send across those logs and console snips.

shouldn't be too hard to sort this out.

**Comment by wellsie** [comment]
oh also... does `systemctl status` on `10.0.0.10` show any issues ?

**Comment by rtlong** [comment]
Thanks for your help! Here's a gist with the logs you wanted: https://gist.github.com/rtlong/2ddd2171c230078c368a

I can't give you the TF output as I've lost it in my shell now, but it was definitely showing the `null_resource.initialize` lines where it described the connection details, followed by (presumable a parallel worker produced this) a block from another resource, then the solitary line, again from `null_resource.initialize` which said merely `Connecting!` and that's the last thing I see before I killed it after 15 minutes or so. (Also possibly of note: I had to ^C twice to break out of that. The first time TF shows it's trying to shutdown cleanly, but it never happens. I assume it's waiting for all active provisioners to finish, so that would make sense.)

**Comment by wellsie** [comment]
np :)

hey remove this (or edit them) from the `cloud-config.tf` files in the worker and etcd modules:

```
        - name: awslogs.conf
          content: |
            [Service]
            Environment="DOCKER_OPTS=--log-driver=awslogs --log-opt awslogs-region=us-west-1 --log-opt awslogs-group=${ log-group }"
```

i have hard coded the region - DOH!!! definitely a bug !

anyway remove them and it should work but won't log to cloudwatch. if you are using a different region try editing to that.

here are the offending log entries:

```
Mar 23 22:45:59 ip-10-0-0-10.ec2.internal kubelet[1132]: E0323 22:45:59.416603    1132 manager.go:1864] Failed to create pod infra container: API error (500): Cannot start container fac40eae07a74838b27fcd0246a69315df15637c5f79205254780850712d2658: Failed to initialize logging driver: ResourceNotFoundException: The specified log group does not exist.
Mar 23 22:45:59 ip-10-0-0-10.ec2.internal kubelet[1132]: status code: 400, request id: []
Mar 23 22:45:59 ip-10-0-0-10.ec2.internal kubelet[1132]: ; Skipping pod "kube-proxy-ip-10-0-0-10.ec2.internal_kube-system"
Mar 23 22:45:59 ip-10-0-0-10.ec2.internal kubelet[1132]: E0323 22:45:59.416678    1132 pod_workers.go:112] Error syncing pod e153596e1cb9c8b7548d10d172f3beea, skipping: API error (500): Cannot start container fac40eae07a74838b27fcd0246a69315df15637c5f79205254780850712d2658: Failed to initialize logging driver: ResourceNotFoundException: The specified log group does not exist.
```

can you check that terraform created a log group for the intended region ?

also delete your gist in case there is anything sensitive in there.

i'll fix this properly tomorrow.

lmk how it goes

**Comment by rtlong** [comment]
That did the trick! Thanks. I'm so glad I reached out. I usually like to spend more time trying to understand something like this before asking for help, but that would have taken me a while to find, I'd guess. I was way off on the wrong track.

**Comment by wellsie** [comment]
Np anytime.

**Comment by rtlong** [comment]
FYI there's also a hard-coded region in `init-cfssl`. Now that I've fixed that as well, I was able to get this to complete, and now have a happy `kubectl` telling me my k8s cluster is up and running. Thanks again for sharing this with the world. I'm trying to get going with the latest pre-release of Deis, so needed to get a k8s cluster going. This has been a helpful jump start, as I know I'm going to us TF/CoreOS on AWS anyway.
